pool/app/filter/repo/repo_utils.ml

```diff
@@ -0,0 +1,223 @@
+module Dynparam = Utils.Database.Dynparam
+open Entity
+
+let where_prefix str = Format.asprintf "WHERE %s" str
+
+let filtered_base_condition =
+  {sql|
+    user_users.admin = 0
+    AND user_users.confirmed = 1
+    AND pool_contacts.paused = 0
+    AND pool_contacts.disabled = 0
+    AND NOT EXISTS
+      (SELECT 1
+      FROM pool_invitations
+      WHERE
+          pool_invitations.contact_id = pool_contacts.id
+        AND
+          pool_invitations.experiment_id IN (
+            SELECT id FROM pool_experiments WHERE pool_experiments.uuid = UNHEX(REPLACE(?, '-', ''))
+            )
+      )
+    AND NOT EXISTS
+      (SELECT 1
+      FROM pool_assignments
+      WHERE
+          pool_assignments.contact_id = pool_contacts.id
+        AND
+          pool_assignments.session_id IN (
+            SELECT id FROM pool_sessions WHERE pool_sessions.experiment_uuid = UNHEX(REPLACE(?, '-', ''))
+          )
+      )
+    |sql}
+;;
+
+let filter_to_sql template_list dyn query =
+  let open Entity in
+  let wrap_in_percentage operator value =
+    let wrap s = CCString.concat "" [ "%"; s; "%" ] in
+    let open Operator in
+    match operator with
+    | ContainsSome | ContainsNone | ContainsAll | Like -> wrap value
+    | Less | LessEqual | Greater | GreaterEqual | Equal | NotEqual -> value
+  in
+  let add_value_to_params operator value dyn =
+    let add c v = Dynparam.add c v dyn in
+    match value with
+    | Bool b -> add Caqti_type.bool b
+    | Date d -> add Caqti_type.ptime d
+    | Language lang -> add Caqti_type.string (Pool_common.Language.show lang)
+    | Nr n -> add Caqti_type.float n
+    | Option id ->
+      add
+        Caqti_type.string
+        (id |> Custom_field.SelectOption.Id.value |> wrap_in_percentage operator)
+    | Str s -> add Caqti_type.string (wrap_in_percentage operator s)
+  in
+  let open CCResult in
+  let rec query_sql (dyn, sql) query : (Dynparam.t * 'a, 'b) result =
+    let of_list (dyn, sql) queries operator =
+      let query =
+        CCList.fold_left
+          (fun res query ->
+            res
+            >>= fun (dyn, lst_sql) ->
+            query_sql (dyn, sql) query
+            >|= fun (dyn, new_sql) -> dyn, lst_sql @ [ new_sql ])
+          (Ok (dyn, []))
+          queries
+      in
+      query
+      >|= fun (dyn, lst_sql) ->
+      ( dyn
+      , lst_sql
+        |> CCString.concat (Format.asprintf " %s " operator)
+        |> Format.asprintf "%s (%s)" sql )
+    in
+    match query with
+    | And queries ->
+      if CCList.is_empty queries
+      then Ok (dyn, sql)
+      else of_list (dyn, sql) queries "AND"
+    | Or queries ->
```

Comment:
wouldn't it be more readable using the "match with" with "when" statements

```
    | Or queries when CCList.is_empty queries ->
    ...
    | Or queries ->
    ...
```

Reply:
👍